added Publishing-Messages-Using-YAML-Files.rst and edited Intermediat… (#5034)
ros2/ros2_documentation: source/Tutorials/Intermediate/Publishing-Messages-Using-YAML-Files.rst

Publishing messages using YAML files
====================================

> **Goal:** Store and publish topics using YAML files.

Reviewer: just a suggestion, i think those words are more familiar with ROS users.

```suggestion
**Goal:** Record and replay topics using YAML files.
```

**Tutorial level:** Intermediate

**Time:** 5 minutes

.. contents:: Table of Contents
   :depth: 2
   :local:

Overview
--------

> Publishing ROS 2 messages via CLI is straightforward for simple types like `std_msgs/msg/Bool <http://docs.ros.org/en/noetic/api/std_msgs/html/msg/Bool.html>`_ or `std_msgs/msg/String <http://docs.ros.org/en/melodic/api/std_msgs/html/msg/String.html>`_.

Reviewer 2: I know the ROS 2 docs/API docs aren't quite stable yet, but I would try to avoid linking to ROS 1 Noetic docs.

I was able to find similar pages for ROS 2, but the URLs aren't quite the same across distros (Humble, Jazzy, Rolling), so we can't just use the same exact link with [`{DISTRO}`](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Contributing-To-ROS-2-Documentation.html#macros) across distros:

* Rolling
    * https://docs.ros.org/en/rolling/p/std_msgs/interfaces/msg/Bool.html
    * https://docs.ros.org/en/rolling/p/std_msgs/interfaces/msg/String.html
* Jazzy
    * https://docs.ros.org/en/jazzy/p/std_msgs/msg/Bool.html
    * https://docs.ros.org/en/jazzy/p/std_msgs/msg/String.html
* Humble
    * https://docs.ros.org/en/humble/p/std_msgs/interfaces/msg/Bool.html
    * https://docs.ros.org/en/humble/p/std_msgs/interfaces/msg/String.html

I would just use the Rolling links with `{DISTRO}` for now:

* https://docs.ros.org/en/{DISTRO}/p/std_msgs/interfaces/msg/Bool.html
* https://docs.ros.org/en/{DISTRO}/p/std_msgs/interfaces/msg/String.html

```suggestion
Publishing ROS 2 messages via CLI is straightforward for simple types like `std_msgs/msg/Bool <https://docs.ros.org/en/{DISTRO}/p/std_msgs/interfaces/msg/Bool.html>`_ or `std_msgs/msg/String <https://docs.ros.org/en/{DISTRO}/p/std_msgs/interfaces/msg/String.html>`_.
```

Then, when backporting to the Humble and Jazzy docs, we can adjust the links.

Author: Makes total sense. I wasn't aware that there is a ROS 2 specific documentation for message definitions.

However, it becomes tedious for complex message structures.

> This tutorial demonstrates how to use YAML files to store, edit, and publish topic data efficiently.

Reviewer: How about the following? it sounds like straight-forward to me what is explained here.

```suggestion
This tutorial demonstrates how to use ``ros2 echo`` and ``ros2 pub`` with YAML files to record, edit, and replay topic data efficiently.
```

Prerequisites
-------------

> This tutorial uses concepts like ROS2 topics and CLI tools covered in the following tutorial:

Reviewer 3: ```suggestion
This tutorial uses concepts like ROS 2 topics and CLI tools covered in the following tutorial:
```

- :doc:`Understanding topics <../Beginner-CLI-Tools/Understanding-ROS2-Topics/Understanding-ROS2-Topics>`

Tasks
-----

We assume that an entity is publishing a ``geometry_msgs/msg/Twist`` message through a topic named ``cmd_vel`` and we want to capture the message, edit it and publish it to a topic.

> We can use the ``echo`` verb to capture the message once using the ``--once`` flag and save it in a YAML file ``cmd_vel.yaml`` using the output redirection operator ``>``.

Reviewer: I am not sure if this `--once` explanation required to put here, because multiple messages already cover that case?

Author: ```suggestion
We can use the ``echo`` verb to capture the message and save it in a YAML file ``cmd_vel.yaml`` using the output redirection operator ``>``.
```
Like that?

Reviewer: @NickTziaros yeah that sounds good to me. thanks

.. code-block:: console

    ros2 topic echo --once  /cmd_vel > cmd_vel.yaml

> This creates a cmd_vel.yaml file with the following content in the directory the command was executed:

Reviewer 3: ```suggestion
This creates a ``cmd_vel.yaml`` file with the following content in the directory the command was executed:
```

.. code-block:: yaml

    linear:
        x: 1.0
        y: 0.0
        z: 0.0
    angular:
        x: 0.0
        y: 0.0
        z: 0.0
    ---

To publish a message, we utilize the ``--yaml-file`` option available with the ``pub`` verb of the ``ros2 topic`` command.

> First, we specify the target topic—in this case,/cmd_vel, followed by the message type ``geometry_msgs/msg/Twist``.

Reviewer 3: ```suggestion
First, we specify the target topic—in this case, ``/cmd_vel``, followed by the message type ``geometry_msgs/msg/Twist``.
```
